File: cpp/set-request-metadata/src/SetRequestMetadata.cpp
/*
 * (c) 2018 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <stdio.h>
#include <cstdlib>
#include <iostream>

int main(int argc, char* argv[]) {

    int nErrorCode;
    std::string keyId = "HVzG5eMZrc8";

    // initialize agent with password persistor
    ISAgent agent;
    ISAgentDeviceProfilePersistorPassword persistor;
    persistor.setFilePath(std::string(std::getenv("HOME")) + "/.ionicsecurity/profiles.pw");
    persistor.setPassword(std::string(std::getenv("IONIC_PERSISTOR_PASSWORD")));
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Error initializing agent: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    // define request metadata
    ISAgentGetKeysRequest request;
    request.getMetadata()["application-state"] = "active";

    // fetch key with request metadata
    request.getKeyIds().push_back(keyId);
    ISAgentGetKeysResponse response;
    nErrorCode = agent.getKeys(request, response);
    ISAgentGetKeysResponse::Key responseKey = response.getKeys().at(0);

    // display fetched key
    ISCryptoHexString hexKey;
    hexKey.fromBytes(responseKey.getKey());
    std::cout << "KeyId    : " << responseKey.getId() << std::endl;
    std::cout << "KeyBytes : " << hexKey << std::endl;
}

File: javascript/set-request-metadata/setRequestMetadata.js
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

/*
 * WARNING *
 * Calling agent.enrollUser() successfully is a prerequisite before using this code.
 * This is done using enrollDevice.js.
 */

import {getAgentConfig} from '../jssdkConfig.js';

const main = async () => {

  const appData = getAgentConfig('JavaScript Set Request Metadata');

  // initialize agent
  try {
    const resp = await new window.IonicSdk.ISAgent(appData);
    const agent = resp.agent;

    // define request metadata
    const appStatus = 'active';
    const appLocation = 'Gondwanaland';

    const requestMetadata = {
      'application-status': appStatus,
      'application-location': appLocation
    };
  
    console.log('');
    let keyId = '';

    // create key on behalf of
    try {
      const response = await agent.createKeys({
        quantity: 1,
        metadata: requestMetadata
      });

      const key = response.keys[0];

      // display created key
      console.log('New Key with key ID: ' + key.keyId);
      console.log('  with application-status = ' + appStatus + ' and application-location = ' + appLocation);
      console.log('KeyBytes          : ' + key.key);
      console.log('FixedAttributes   : ' + JSON.stringify(key.attributes,null,0));
      console.log('MutableAttributes : ' + JSON.stringify(key.mutableAttributes,null,0));

      keyId = key.keyId;
    } catch (sdkErrorResponse) {
        console.log('Error Creating Key: ' + sdkErrorResponse.error);
    }

    // get key with request metadata
    console.log(' ');
    try {
      const response = await agent.getKeys({
        keyIds: [keyId],
        metadata: requestMetadata
      });

      const key = response.keys[0];
    
      // display fetched key
      console.log('Getting key with application-status = ' + appStatus + ' and application-location = ' + appLocation);
      console.log('KeyId             : ' + key.keyId);
      console.log('KeyBytes          : ' + key.key);
      console.log('FixedAttributes   : ' + JSON.stringify(key.attributes, null, 0));
      console.log('MutableAttributes : ' + JSON.stringify(key.mutableAttributes, null, 0));
    } catch (sdkErrorResponse) {
        console.log('Error Creating Key: ' + sdkErrorResponse.error);
    }
  } catch (sdkErrorResponse) {
    console.log('Initializing agent error: ' + sdkErrorResponse.error);
  }
}

main()
